_posts/2024-05-22-developing-openwrt-in-qemu.adoc

```diff
@@ -0,0 +1,93 @@
+---
+title: Developing OpenWRT in QEMU with Eclipse Che
+layout: post
+author: Oleksii Orel
+description: >-
+  Developing OpenWRT helloworld package in QEMU machine emulator and virtualizer with Eclipse Che.
+categories: []
+keywords: ['openwrt', 'quemu', 'gdb', 'developing ', 'workspace']
+slug: /@olexii.orel/openwrt-helloworld-package
+---
+
+https://openwrt.org/[OpenWrt] is a highly extensible GNU/Linux distribution for embedded devices (typically wireless routers). OpenWrt is built from the ground up to be a full-featured, easily modifiable operating system for embedded devices. You can have all the features you need with no bloat, powered by a modern Linux kernel.
+
+We can create and develop https://openwrt.org/docs/guide-developer/creating-a-cmake-package-in-openwrt[OpenWRT CMake packages] with Eclipse Che.
+
+In this blog post, we will review developing OpenWRT in https://www.qemu.org/[QEMU] with Eclipse Che in 4 parts:
+
+<<part1>>
+
+<<part2>>
+
+<<part3>>
+
+<<part4>>
+
+
+NOTE: In this blog post, we will use https://github.com/che-incubator/openwrt-helloworld-package.git[OpenWrt HelloWorld package].
+
+
+First of all, we need to create a new workspace from the repo https://github.com/che-incubator/openwrt-helloworld-package.git[openwrt-helloworld-package] with Eclipse Che. The easiest way to do it is to use https://eclipse.dev/che/docs/stable/hosted-che/hosted-che/[Eclipse Che hosted by Red Hat] and create a workspace by navigating to the following URL - https://workspaces.openshift.com#https://github.com/che-incubator/openwrt-helloworld-package/blob/main/devfile.yaml
+
+
+## PART 1: Quick start with prebuild recourses [[part1]]
+
+Prebuilt resources could be applied for a quick start with Eclipse Che. We could do it by running the next tasks from the https://github.com/che-incubator/openwrt-helloworld-package/blob/main/devfile.yaml[devfile]:
+
+1. `Add prebuilt resources`
+
+NOTE: After adding prebuilt resources, we should see the `openwrt-*.img.gz` file in the `output` directory. Prebuilt resources were added to the runtime image during the last successful GitHub image build action.
+
+image::/assets/img/developing-openwrt/add-prebuilt-resources.png[Add prebuilt resources]
+{nbsp} +
+
+## PART 2: (optinal)Build all packages and the kernel [[part2]]
+
+After creating a new workspace from https://github.com/che-incubator/openwrt-helloworld-package.git[openwrt-helloworld-package] we need to install all package definitions, copy https://github.com/che-incubator/openwrt-helloworld-package/blob/main/configs/.x86-generic.config[`.config`] file and build all packages and the kernel. We can do it by running the following tasks from the https://github.com/che-incubator/openwrt-helloworld-package/blob/main/devfile.yaml[devfile]:
```

Comment:
```suggestion
Instead of using prebuilt resources from STEP 1, we can install all package definitions, copy https://github.com/che-incubator/openwrt-helloworld-package/blob/main/configs/.x86-generic.config[`.config`] file and build all packages and the kernel. We can do this by running the following tasks from the https://github.com/che-incubator/openwrt-helloworld-package/blob/main/devfile.yaml[devfile]:
```
I suggest changing `need` to `can` if this is an optional step